feat(chips): add support for custom separator keys

Ignigena opened the pull request:
Implements #5279 

Add the optional ability for chips to be created on keydown of any key
in the `md-separator-keys` attribute. Custom key codes are supported
in addition to common ones defined in `$mdConstant.KEY_CODE`.

Ignigena commented:
I signed it!

Frank3K commented:
There's also a valid use case for a semicolon-separator, e.g. when entering multiple email-addresses. I think it would be more flexible if one could specify a list of separators instead of having an attribute specifically for the comma-separated option.

What do you think?

Ignigena commented:
@Frank3K Good point. And I'm definitely all for a more generic option to support for other use cases. Let me see what I can do...

topherfangio commented:
This looks really interesting. Can you add/edit a demo to show how this would work?

Ignigena commented:
Here's a quick and dirty demo: http://codepen.io/anon/pen/rOKexW

Top demo is comma separated and bottom demo adds the ability to separate on semi-colons by injecting and modifying `$mdConstant.KEY_CODES`

topherfangio commented:
@Ignigena That demo looks great, but I was hoping you could incorporate it into an existing demo (or create a new one) so that end-users would have something to look at as an example :smile:

Ignigena commented:
Ahhh I see now. Of course! Sorry, it's been a long day :smile:

Ignigena commented:
Basic demo for both comma and custom separators pushed. I incorporated it into a new demo so as not to complicate the existing "basic usage" one further.

Frank3K commented:
Is it good practice to add keys to the $mdConstant object by changing the object after creation? Is does not completely feel right to me.

topherfangio commented:
@Ignigena - @Frank3K makes a good point. Rather than using strings for the keys, I think it would be better if you passed the actual key codes so that you could pass constants defined in `$mdConstant` or some service provided by your own application.

Forcing users to inject the key codes into `$mdConstant` seems like it could cause trouble later on.

Ignigena commented:
@topherfangio Makes complete sense. Refactoring now.

Ignigena commented:
Refactored and updated demo and spec. I left the addition of the `COMMA` definition on the constant since it seems there would be a broad enough use case for this, but let me know if this should be removed in favor of a strictly generic approach.

ThomasBurleson commented:
@topherfangio - is the ready for merging ?

topherfangio commented:
@ThomasBurleson LGTM.

DevVersion commented:
I love that feature. Great job @Ignigena

topherfangio commented:
@Ignigena Can you rebase against the latest master branch and squash your commits into one so that @ThomasBurleson can merge when he is ready?

Thanks!

Ignigena commented:
@topherfangio Done and done!

topherfangio commented:
Awesome, thanks so much!

ThomasBurleson commented:
@Ignigena - thank you for your PR and your responsiveness to our change requests.
